### 셀레니움 : **pyautogui/화면 캡처**

In [ ]:
!pip install pyautogui
# 외부 라이브러리라서 설치 먼저 해야 함!

In [ ]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time
import pyautogui # >>>>>>>> NEW!!

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu") 
options.add_argument("--start-maximized") 

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://davelee-fun.github.io/")

#
pyautogui.size() #반환값으로 너비,높이 찾아와줌
# print(pyautogui.size(),"\n") => Size(width=1920, height=1080)

screen_width, screen_height = pyautogui.size() # 첫 번째 인자값을 첫 번째 변수에, 두 번째 인자값을 두 번째 변수에 넣어줌
# print(screen_width, screen_height) =>1920 1080

driver.set_window_size(screen_width,screen_height) # (가로(너비),세로(높이))
element = driver.find_element(By.TAG_NAME, "body")

element.screenshot("samplesite.png")
# .screenshot : 해당 요소의 해당 브라우저 화면을 캡처해 와라!+"samplesite.png"라는 이름으로 => 작업 중인 루트 폴더(250723)에 저장됨

time.sleep(3)
driver.quit()

### 📌 **Headless** 기능
- 실제 브라우저를 띄우지 않아도 마치 띄운 것처럼 인식시켜 주게 하는 기능
- 사이트 n초 동안 뜨는 거 없애 줌

In [43]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--start-maximized")
options.add_argument("--headless") #>>>>> new!! (창 뜨는 거 막아줌->그만큼 해당 사이트는 민감하게 반응하니 주의!)
# headless까지도 속이기 위해 아래 옵션을 더 넣어줌
options.add_argument("--window-size=1920x1080")
# options.add_argument("User-Agent: Mozilla/5.0 (Linux; Android 6.0; Chrome 5...") 너무 많음!! 이걸 일일히 쓸 수 없어서 찾아야 함
# 웹사이트 -> console -> window.navigator.userAgent : 내 크롬 브라우저의 정보값
# 결과 : 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36' 붙여넣기
options.add_argument("--User-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36")
options.add_argument("--lang=ko_KR")
#이렇게 추가하면 완벽하게 속음

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://davelee-fun.github.io/")

element = driver.find_element(By.TAG_NAME, "h1")
print(element.text)

time.sleep(1)

driver.quit()

Teddy의 선물 블로그


### 🚩 **MISSION** 다음 웹 사이트에 접속해서 selenium을 활용하여 크롤링 해 오시오. -1
- https://crawlingtest.netlify.app/
- 부모 요소의 id 속성값이 dev_course_list이면서 class가 course인 데이터만 출력하시오.

In [4]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--start-maximized")
options.add_argument("--headless")
options.add_argument("--window-size=1920x1080")
options.add_argument("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36")
options.add_argument("--lang=ko_KR")

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://crawlingtest.netlify.app/")

elements = driver.find_elements(By.CSS_SELECTOR, "ul#dev_course_list > li.course")

for element in elements:
    print(element.text)

time.sleep(3)
driver.quit()

(초급) - 강사가 실제 사용하는 자동 프로그램 소개 [2]
(초급) - 필요한 프로그램 설치 시연 [5]
(초급) - 데이터를 엑셀 파일로 만들기 [9]
(초급) -     엑셀 파일 이쁘게! 이쁘게! [8]
(초급) -     나대신 주기적으로 파이썬 프로그램 실행하기 [7]
(초급) - 파이썬으로 슬랙(slack) 메신저에 글쓰기 [40]
(초급) - 웹사이트 변경사항 주기적으로 체크해서, 메신저로 알람주기 [12]
(초급) - 네이버 API 사용해서, 블로그에 글쓰기 [42]
(중급) - 자동으로 쿠팡파트너스 API 로 가져온 상품 정보, 네이버 블로그/트위터에 홍보하기 [412]


### 🚩 **MISSION** 다음 웹 사이트에 접속해서 selenium을 활용하여 크롤링 해 오시오. -2
- https://crawlingtest.netlify.app/
- 해당 사이트 내 커리큘럼 일정표 안에 있는 데이터를 모두 추출해서 출력하시오.

In [8]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--start-maximized")
options.add_argument("--headless")
options.add_argument("--window-size=1920x1080")
options.add_argument("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36")
options.add_argument("--lang=ko_KR")

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://crawlingtest.netlify.app/")

elements = driver.find_elements(By.CSS_SELECTOR, "td") #정석대로는 (By.CSS_SELECTOR, "table tr td")
# 같은 답 = elements = driver.find_elements(By.TAG_NAME, "td")

for element in elements:
    print(element.text)

time.sleep(3)
driver.quit()

일정
커리큘럼 타이틀
난이도
5.1 ~ 6.15
나만의 엣지있는 블로그 사이트 만들기 (취미로 익히는 IT)
초급
6.16 ~ 7.31
파이썬과 데이터과학 첫걸음 (IT 기본기 익히기)
중급


### 🚩 **MISSION** 다음 웹 사이트에 접속해서 selenium을 활용하여 크롤링 해 오시오. -3
- https://crawlingtest.netlify.app/
- 아래 데이터만 추출해 오시오
- (중급) - 자동으로 쿠팡파트너스 API 로 가져온 상품 정보, 네이버 블로그/트위터에 홍보하기 [412]

In [26]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--start-maximized")
options.add_argument("--headless")
options.add_argument("--window-size=1920x1080")
options.add_argument("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36")
options.add_argument("--lang=ko_KR")

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://crawlingtest.netlify.app/")

elements = driver.find_elements(By.CSS_SELECTOR, "ul#dev_course_list > li.course")
print(elements[-1].text) # 답 하나만 추출할 것 >>>>> 리스트로 뽑을 게 아니니까 for문 안 써도 된다!!!
# 같은 답 = print(elements[8].text) 

time.sleep(1)
driver.quit()

(중급) - 자동으로 쿠팡파트너스 API 로 가져온 상품 정보, 네이버 블로그/트위터에 홍보하기 [412]


### 🚩 **MISSION** 다음 웹 사이트에 접속해서 selenium을 활용하여 크롤링 해 오시오. -4
- https://divjason.github.io/sellenium-test/
- class값이 "news"인 데이터를 출력하시오.
- span 태그의 news 클래스 !!!!

In [33]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--start-maximized")
options.add_argument("--headless")
options.add_argument("--window-size=1920x1080")
options.add_argument("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36")
options.add_argument("--lang=ko_KR")

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://divjason.github.io/sellenium-test/")

element = driver.find_element(By.CSS_SELECTOR, "span.news") # 굳이 앞에 section.main 안 적어도 답 잘 나온다!
print(element.text)

time.sleep(1)
driver.quit()

뉴스: 머신러닝 강좌가 나왔어요~~~


### 🚩 **MISSION 5-1** 다음 웹 사이트에 접속해서 selenium을 활용하여 크롤링 해 오시오.
- https://divjason.github.io/sellenium-test/
- input 요소에 아이디 error@error.com / 비밀번호 1234 입력 후, 로그인 버튼 자동으로 클릭하기
### 🚩 **MISSION 5-2**
- 클릭 이후 이동한 페이지에서 message라는 클래스 속성값을 가지고 있는 요소의 텍스트 정보를 출력하기
### 🚩 **MISSION 5-3**
- 로그아웃 버튼 클릭 후 이동한 페이지에서 다시 message라는 클래스 속성값을 가지고 있는 요소의 텍스트 정보를 출력하기

In [78]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--start-maximized")

# 창이 안 떠서 headless 잠깐 주석 처리!
# options.add_argument("--headless")
# options.add_argument("--window-size=1920x1080")
# options.add_argument("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36")
# options.add_argument("--lang=ko_KR")

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://divjason.github.io/sellenium-test/")

ID = driver.find_element(By.NAME, "username")   # By.ID,"username"도 가능, By.CSS_SELETOR도 가능)
ID.clear() # 항상 초기화 먼저!!!
ID.send_keys("error@error.com")

PW = driver.find_element(By.NAME, "password")
PW.clear()
PW.send_keys("아아6")

ID.send_keys(Keys.RETURN)

time.sleep(1) # 이 머무르는 시간이 있어야 데이터 나옴

element01 = driver.find_element(By.CSS_SELECTOR, "div.message")
print(element01.text)

time.sleep(1)

button = driver.find_element(By.CSS_SELECTOR, "input[type='submit']")
button.click

time.sleep(1)

element02 = driver.find_element(By.CSS_SELECTOR, "div.message")
print(element02.text)

time.sleep(3)
driver.quit()

실제 사이트는 https://megaitlab.com/ 입니다
본 페이지는 로그인된 페이지입니다!
뉴스: 머신러닝 강좌가 나왔어요~~~
실제 사이트는 https://megaitlab.com/ 입니다
본 페이지는 로그인된 페이지입니다!
뉴스: 머신러닝 강좌가 나왔어요~~~


In [80]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--start-maximized")

# 창이 안 떠서 headless 잠깐 주석 처리!
# options.add_argument("--headless")
# options.add_argument("--window-size=1920x1080")
# options.add_argument("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36")
# options.add_argument("--lang=ko_KR")

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://divjason.github.io/sellenium-test/")

ID = driver.find_element(By.NAME, "username")   # By.ID,"username"도 가능, By.CSS_SELETOR도 가능)
ID.clear() # 항상 초기화 먼저!!!
ID.send_keys("error@error.com")

PW = driver.find_element(By.NAME, "password")
PW.clear()
PW.send_keys("아아6")

ID.send_keys(Keys.RETURN)

time.sleep(1) # 이 머무르는 시간이 있어야 데이터 나옴

element01 = driver.find_element(By.CSS_SELECTOR, "div.message")
print(element01.text)

time.sleep(3)
driver.quit()

실제 사이트는 https://megaitlab.com/ 입니다
본 페이지는 로그인된 페이지입니다!
뉴스: 머신러닝 강좌가 나왔어요~~~


엔터 (Keys.RETURN) 말고 정석 (By.CSS_SELECTOR, "input[type='submit']")

In [81]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

service = Service("C:\\Users\\user\\Downloads\\chromedriver\\chromedriver.exe")
options = Options()
options.add_argument("--disable-gpu")
options.add_argument("--start-maximized")

# 창이 안 떠서 headless 잠깐 주석 처리!
# options.add_argument("--headless")
# options.add_argument("--window-size=1920x1080")
# options.add_argument("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36")
# options.add_argument("--lang=ko_KR")

driver = webdriver.Chrome(service=service, options=options)
driver.get("https://divjason.github.io/sellenium-test/")

ID = driver.find_element(By.NAME, "username")   # By.ID,"username"도 가능, By.CSS_SELETOR도 가능)
ID.clear() # 항상 초기화 먼저!!!
ID.send_keys("error@error.com")

PW = driver.find_element(By.NAME, "password")
PW.clear()
PW.send_keys("12345")

time.sleep(5) # ID,PW가 제대로 쳐졌는지 확인용 타임슬립!

#ID.send_keys(Keys.RETURN) -> 주석처리!
#================
# input 태그 애초에 ID/PW -> **엔터** -> 알아서 서버로 전송 (기본 개념)
# 엔터키가 안 먹게 할 수도 있음!!! 이유 : 버튼 클릭(ex.로그인버튼)에 대한 이벤트를 설정할 때!(엔터로 서버 전송된 정보는 수집이 안 돼서 클릭율을 알아볼 수 없음ㅠ)
# => prevent default -> 방해하는 기능
# 예를 들어 둘 중 하나 선택하는 이벤트에서는 꼭 필요함! 구체적으로 버튼을 클릭할 수밖에 없게끔!

element = driver.find_element(By.CSS_SELECTOR, "input[type='submit']")
#CSS_SELECTOR에서 속성값을 찾아올 때 이런 형태로! 'submit' 이 속성값을 큰따옴표로 쓰면 오류나기 딱 좋아서 꼭 작은따옴표 안에 넣어 줘야 함!
# html CRM 에 대한 정리가 하고 싶을 때! MDN 가서 검색!
element.click()

print("🎉 로그인 성공!")

time.sleep(5)
driver.quit()

🎉 로그인 성공!
